Add support for DNSLink resolution

spylogsster opened the pull request:
Added an Open in IPFS option for domains that have a TXT record that is prefixed with dnslink=

<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/13609
Resolves https://github.com/brave/brave-browser/issues/13611

## Submitter Checklist:

- [x] Requested security review https://github.com/brave/security/issues/332
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Navigate to brantly.eth.link and en.wikipedia-on-ipfs.org, on both you should see Open in IPFS option in address bar only if IPFS mode is local or gateway
- Click by button will open ipns://[domain] url in new tab, if you have any path in url, it should be added as well
- This should work for all domains that have dnslink= in the TXT dns record and  x-ipfs-path HTTP header is present
- This should work if page returned code different from OK 200 or failed due to network/DNS
- This must not work in private windows or in Tor windows
- Please note it doesnt work for me of I use Secure DNS "with your current service provider" as there are no available dns serviers. It may be same on your side

diracdeltas commented on browser/ipfs/ipfs_host_resolver.cc:
```diff
@@ -0,0 +1,94 @@
+
+namespace {
+const uint8_t kGooglePublicDns1[] = {8, 8, 8, 8};
+const uint8_t kGooglePublicDns2[] = {8, 8, 4, 4};
```
are we allowed to use this without users agreeing to the google DNS terms of service? https://developers.google.com/speed/public-dns/terms

spylogsster replied:
it is already used in chromium here https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/net/dns_probe_service_factory.cc;l=292;drc=6531b4f2fda5515c3315e50d1cfb709e607d53e6;bpv=1;bpt=1, 
I tried to use default DNS it works only for A/AAAA records and reports errors for attempts to read TXT

spylogsster replied:
Removed this code and pushed new one same as in dns_probe to meet actual browser prefs around DoH and secure dns, thanks

diracdeltas commented on browser/ipfs/ipfs_host_resolver.h:
```diff
@@ -0,0 +1,69 @@
+
+  std::string host() const { return resolving_host_; }
+
+  network::mojom::HostResolver* GetHostResover() {
```
nit: should be `GetHostResolver`

diracdeltas commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +105,96 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+
+void IPFSTabHelper::ResolveIPFSLink() {
+  if (web_contents()->GetBrowserContext()->IsTor() ||
+      web_contents()->GetBrowserContext()->IsOffTheRecord())
```
cc @darkdh to verify this is sufficient to disable resolution in Tor

darkdh replied:
This is redundant, first `IsOffTheRecord` already include Tor window. Secondly, `IPFSTabHelper` won't even get created for Tor window, see `IPFSTabHelper::MaybeCreateForWebContents` ->`ipfs::IpfsServiceFactory::GetForContext`->`ipfs::IsIpfsEnabled`.
So please remove this check.

spylogsster replied:
done

diracdeltas commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +105,96 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+  if (ipfs_resolved_host_.empty() &&
+      response_code >= net::HttpStatusCode::HTTP_INTERNAL_SERVER_ERROR &&
+      response_code <= net::HttpStatusCode::HTTP_GATEWAY_TIMEOUT) {
+    ResolveIPFSLink();
```
why is this called in this case?

spylogsster replied:
https://github.com/brave/brave-browser/issues/13609#issuecomment-764627805
```
or if page returned code different from OK 200 or failed due to network/DNS (eg. invalid A record)
(this will make it possible for people to recover from broken/DDoSs HTTP servers etc)
```

diracdeltas commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +105,96 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+  return false;
+}
+
+void IPFSTabHelper::ResolveIPFSLink() {
```
should this return early if the resolve method is IPFSResolveMethodTypes::IPFS_DISABLED ?

diracdeltas commented:
to avoid an extra DNS request (at least for users who haven't yet opted into IPFS), i wonder if it's possible to look up this TXT record in the DNS cache instead? maybe https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_cache.h?q=dns%20cache&ss=chromium ?

spylogsster commented:
> to avoid an extra DNS request (at least for users who haven't yet opted into IPFS), i wonder if it's possible to look up this TXT record in the DNS cache instead? maybe https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_cache.h?q=dns%20cache&ss=chromium ?

This cache should be used inside HostResolver, https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_resolver_manager.h;l=56?q=HostResolverManager&ss=chromium%2Fchromium%2Fsrc,

spylogsster commented on browser/ipfs/ipfs_host_resolver.cc:
```diff
@@ -0,0 +1,112 @@
+    }
+    current_config_overrides.secure_dns_mode = net::SecureDnsMode::kSecure;
+  } else {
+    // A DNS error that occurred in automatic mode must have had an insecure
```
@diracdeltas this is from https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/net/dns_probe_service_factory.cc;l=244
probably same logic should work for our case as well. what do you think?

diracdeltas replied:
i don't think we should do this because the chromium code only runs when there's already been a DNS error. the definition of `kOff` says it will try lookup without DoH as a fallback if there's an error: https://source.chromium.org/chromium/chromium/src/+/master:net/dns/public/secure_dns_mode.h;l=19;drc=04559eae8e59a730ab668d137cdc363ada8c6b24. the code here doesn't assume that there's already a DNS error.

i'm not familiar with this code, but it seems like `current_config_overrides.secure_dns_mode` in this case should just be `secure_dns_config.mode()` ?

diracdeltas commented on browser/ui/views/location_bar/ipfs_location_view.cc:
```diff
@@ -0,0 +1,154 @@
+    Browser* browser = chrome::FindTabbedBrowser(profile_, true);
+    if (!browser)
+      return;
+    content::OpenURLParams open_tor(ipfs_location_, content::Referrer(),
```
should probably call this `open_ipfs` for clarity

diracdeltas commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -39,6 +64,21 @@ void SetupIPFSProtocolHandler(const std::string& protocol) {
 }
+
+// Expects dns TXT record in format: name=value
+std::string GetDSNRecordValue(const std::vector<std::string>& text_results,
```
`GetDNSRecordValue` instead of `GetDSNRecordValue`

diracdeltas commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +110,93 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+    const std::string& host,
+    const std::vector<std::string>& text_results) {
+  std::string dnslink = GetDSNRecordValue(text_results, kDnsLinkHeader);
+  if (dnslink.empty())
```
i think we want to add more validation here before passing the TXT value to IPFS. there are certain URLs that websites are not allowed to open for security reasons, for instance `file:///etc/hosts` or `chrome://settings`. if an attacker creates a site with a dnslink value set to a privileged URL like this, it should not open.

spylogsster commented:
> > to avoid an extra DNS request (at least for users who haven't yet opted into IPFS), i wonder if it's possible to look up this TXT record in the DNS cache instead? maybe https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_cache.h?q=dns%20cache&ss=chromium ?
> 
> This cache should be used inside HostResolver, https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_resolver_manager.h;l=56?q=HostResolverManager&ss=chromium%2Fchromium%2Fsrc,
> 
> Added flag to allow cache usage in parameters and changed code to use one resolver instance per tab helper, but in my tests with Wireshark I see DNS requests for same sites anyway. I will check this logic, probably would be better to create a separate issue for this

@diracdeltas I added patch to Chromium code to use cached TXT results in resolver responses, cache works now

diracdeltas commented on browser/ipfs/ipfs_host_resolver.cc:
```diff
@@ -0,0 +1,87 @@
+}
+
+void IPFSHostResolver::OnTextResults(const std::vector<std::string>& results) {
+  VLOG(1) << results.size()
```
minor: this should probably be at a higher logging level since it's not an error

spylogsster replied:
done

bridiver commented on chromium_src/services/network/resolve_host_request.cc:
```diff
@@ -0,0 +1,13 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#define BRAVE_SINGAL_TEXT_RESULTS                     \
```
don't we get these in `OnComplete` anyway which is called right after this?

bridiver replied:
oh wait, I see, it calls response_client_->OnComplete directly when not pending, but then shouldn't we call `SignalNonAddressResults`  instead of duplicating here?

spylogsster replied:
SignalNonAddressResults calls class member **response_client_,**, but we call local instance of **response_client** it should be moved later [here](https://source.chromium.org/chromium/chromium/src/+/master:services/network/resolve_host_request.cc;l=69?q=resolve_host_request.cc&ss=chromium).
This is only for case when we got data from local dns cache, Chromium works with A/AAAA records and no need to call onTextResults. We are first here and maybe we should push this patch to upstream

bridiver commented on browser/ipfs/ipfs_host_resolver.h:
```diff
@@ -0,0 +1,67 @@
+
+  void Resolve(const net::HostPortPair& host,
+               const net::NetworkIsolationKey& isolation_key,
+               net::DnsQueryType dns_query_type,
```
looks like we're missing an include for `DnsQueryType` and `HostPortPair`

bridiver replied:
a few others below too like base::Optional, etc... make sure to include what you use directly even if it's an override

bridiver commented on browser/ipfs/ipfs_host_resolver.h:
```diff
@@ -0,0 +1,67 @@
+
+  network::mojom::NetworkContext* network_context_ = nullptr;
+  HostTextResultsCallback resolved_callback_;
+  base::OnceClosure reset_callback_for_testing_;
```
also missing an include here - see https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md

bridiver commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -39,6 +65,21 @@ void SetupIPFSProtocolHandler(const std::string& protocol) {
 }
+
+// Expects dns TXT record in format: name=value
+std::string GetDNSRecordValue(const std::vector<std::string>& text_results,
```
is there no existing code to parse TXT records in chromium?

spylogsster replied:
no, it's just a list of strings, each can be in any format

bridiver replied:
ah, right. I forgot it has no specified format

bridiver commented on browser/ipfs/ipfs_host_resolver.cc:
```diff
@@ -0,0 +1,87 @@
+    Reset();
+
+  resolving_host_ = host.host();
+  net::HostPortPair local_host_port(prefix_ + resolving_host_, host.port());
```
shouldn't `_dnslink` be appended to the domain name and not the hostname? For instance if you have `www.whatever.com`, you want `_dnslink.whatever.com`, right? Not `_dnslink.www.whatever.com`?

bridiver replied:
I'm looking at https://docs.ipfs.io/concepts/dnslink/#publish-using-a-subdomain and it's a bit confusing because it talks about domains, not hostnames

bridiver commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -39,6 +65,21 @@ void SetupIPFSProtocolHandler(const std::string& protocol) {
+std::string GetDNSRecordValue(const std::vector<std::string>& text_results,
+                              const std::string& name) {
+  for (const auto& txt : text_results) {
+    if (!base::StartsWith(txt, name))
```
seems like it would be safer to split and then check the equality of the name value. I mean it's unlikely someone would do dnslinkblah=whatever, but I also don't really see any benefit to checking StartsWith before splitting

bridiver commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +111,115 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+  if (current.host() != host || !current.SchemeIsHTTPOrHTTPS())
+    return;
+  std::string dnslink = GetDNSRecordValue(text_results, kDnsLinkHeader);
+  if (dnslink.empty())
```
we don't do anything with this value? I think this needs a comment explaining what we're doing here

spylogsster replied:
yes, we should only check if it exists, https://github.com/brave/brave-browser/issues/13609#issuecomment-790030606
I will add a comment

bridiver commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +111,115 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+
+  int response_code = handle->GetResponseHeaders()->response_code();
+  if (ipfs_resolved_host_.empty() &&
+      response_code >= net::HttpStatusCode::HTTP_INTERNAL_SERVER_ERROR &&
```
can you also add comments here? Are there any ipfs docs you can link to explaining why we're handling this subset of status codes?

spylogsster replied:
https://github.com/brave/brave-browser/issues/13609#issuecomment-764627805
```
or if page returned code different from OK 200 or failed due to network/DNS (eg. invalid A record)
(this will make it possible for people to recover from broken/DDoSs HTTP servers etc)
```

bridiver commented on browser/ui/views/location_bar/brave_location_bar_view.h:
```diff
@@ -8,6 +8,7 @@
 
 #include <vector>
 
+#include "brave/components/ipfs/buildflags/buildflags.h"
```
looks like we're missing a dep for `//brave/components/ipfs/buildflags` in `//brave/browser/ui` (it currently has check_includes = false because of some other unresolved issues)

bridiver commented on browser/ui/views/location_bar/ipfs_location_view.cc:
```diff
@@ -0,0 +1,154 @@
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/layout_constants.h"
+#include "chrome/browser/ui/views/toolbar/toolbar_ink_drop_util.h"
+#include "components/grit/brave_components_resources.h"
```
I think we're missing some deps here, too. This target has `check_includes = false` so you have to manually check unfortunately

bridiver commented on patches/net-dns-host_cache.cc.patch:
```diff
@@ -0,0 +1,12 @@
+     }
+     copy.set_hostnames(std::move(hostnames_with_port));
+   }
++  BRAVE_COPY_TEXT_RECORDS
```
patches should be `BRAVE_ORIGINAL_METHOD_NAME` so `BRAVE_COPY_WITH_DEFAULT_PORT`. The reason is that patches are specifically intended to be flexible and extensible so they shouldn't be named for what they do since that could change or things can be added.

bridiver commented on patches/services-network-resolve_host_request.cc.patch:
```diff
@@ -0,0 +1,12 @@
+   mojo::Remote<mojom::ResolveHostClient> response_client(
+       std::move(pending_response_client));
+   if (rv != net::ERR_IO_PENDING) {
++    BRAVE_SINGAL_TEXT_RESULTS
```
same as above, convention would be `BRAVE_START`

bridiver commented on browser/ipfs/ipfs_host_resolver.cc:
```diff
@@ -0,0 +1,87 @@
+                               HostTextResultsCallback callback) {
+  if (!callback)
+    return;
+  if (IsRunning())
```
It seems like we're adding complexity to this by making it reusable and I'm not sure it's 100% correct. For instance:
1. Call to resolve 1
2. Navigate to another page (callback is invalidated)
3. Call to resolve 2 (callback is now valid again)
4. Response from resolve 1 runs callback with incorrect results
Seems like it would be simpler to just subclass `ResolveHostClientImpl`, implement `OnTextResults` and reset to a new instance on navigation?

spylogsster replied:
IPFSHostResolver is only wrapper over network_context_->ResolveHost which manages resolvers.

I reset reciever on line 42, it will unbound reciever and reset actual resolver inside network service process. Third step will work with new instance each time as you proposed and it is impossible to get case from 4 step now.

I tried subclassing of ResolveHostClientImpl first but it increases necessary amount of code to achieve same result as we have here.

bridiver commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +111,115 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+    content::OpenURLParams params(GetIPFSResolvedURL(), content::Referrer(),
+                                  WindowOpenDisposition::CURRENT_TAB,
+                                  ui::PAGE_TRANSITION_LINK, false);
+    web_contents()->OpenURL(params);
```
what happens if there is a txt record for both the hostname and _dnslink.hostname? It seems like once we get a valid result we should abort any other in-progress requests? Also since the hostname request will be cached it seems like it would be better to only run the _dnslink.hostname query if the hostname query doesn't have a dnslink record?

spylogsster replied:
we do not know what domain will have TXTrecord with dnslink, so we have to check both all time. Once we ignore dnslink value and should only check for presence of this TXT record it doenst matter for us which one will be finished first.

bridiver commented on browser/ipfs/ipfs_tab_helper.cc:
```diff
@@ -62,19 +111,115 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+  if (ipfs_resolved_host_.empty() &&
+      response_code >= net::HttpStatusCode::HTTP_INTERNAL_SERVER_ERROR &&
+      response_code <= net::HttpStatusCode::HTTP_GATEWAY_TIMEOUT) {
+    ResolveIPFSLink();
```
won't this cause us to make the ipfs dns calls for any site that returns one of these error codes? Seems like we should have `if (!handle->GetResponseHeaders()->HasHeader("x-ipfs-path")) return;` at the top of this method?

bridiver replied:
we should probably have some browser tests to ensure this stuff is only called for sites with the header
